### ISSUE TITLE
Router context invalidate Promise resolve does not invalidate context

### ISSUE THREAD

**Issue by ZahidHasan321** [TanStack/router#2072]
### Describe the bug

In the code below, The code was suppose to invalidate the context and navigate to the following route. for context i'm following the example given in tanstack authenticated routes example. Even after navigation the context isn't invalidated which I can see in console log. It results in redirect to login page not working properly after clicking logout and just stuck in the "/" page.  
`router.invalidate().finally(() => {
          navigate({ to: "/" });
        });`
        
  PS: I do have an api request in the logout function.

### Your Example Website or App

...

### Steps to Reproduce the Bug or Issue

Following the tanstack example locally.

### Expected behavior

Invalidate router context..

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by robertyulisman** [comment]
i got same issue, and stuck with this, any solution ? when i am use authenticated using local storage, there is no problem, but when using api, not navigate after login success

**Comment by ZahidHasan321** [comment]
> i got same issue, and stuck with this, any solution ? when i am use authenticated using local storage, there is no problem, but when using api, not navigate after login success

Still no reply from the devs as of now

**Comment by ilia-luk** [comment]
same issue here

**Comment by schiller-manuel** [comment]
can you please provide a minimal complete example by forking one of the router examples on stackblitz?

**Comment by ZahidHasan321** [comment]
> can you please provide a minimal complete example by forking one of the router examples on stackblitz?

https://stackblitz.com/edit/tanstack-router-4achrw?file=src%2Froutes%2F_auth.tsx 

Here a link to a fork, I've only made one change to this, instead of navigating to "/" which is a public route, it navigates to "login", which shouldn't be a problem if context is invalidated, however it isn't. so even after clicking logout, it doesn't go back to "login" route.

**Comment by mkhennoussi** [comment]
Same here while using supabase auth api. No issue when working with local storage.

**Comment by mkhennoussi** [comment]
update guys, the current hack seems to use `sleep(1000)` as explained here : 
https://github.com/TanStack/router/discussions/1668

```js
// This is just a hack being used to wait for the auth state to update
// in a real app, you'd want to use a more robust solution
      await sleep(1)
```

This trick works on my side, but this is a serious issue with Context.

**Comment by ilia-luk** [comment]
I'm currently using the [flushSync](https://react.dev/reference/react-dom/flushSync) solution as a workaround following suggestion by SzelamC at https://github.com/TanStack/router/discussions/1668#discussioncomment-9726727

**Comment by ZahidHasan321** [comment]
> I'm currently using the [flushSync](https://react.dev/reference/react-dom/flushSync) solution as a workaround following suggestion by SzelamC at [#1668 (reply in thread)](https://github.com/TanStack/router/discussions/1668#discussioncomment-9726727)

So where should I use flushSync regarding this issue?

**Comment by mkhennoussi** [comment]
Thanks @ilia-luk, it works great, I didn't thought about this solution. 

@ZahidHasan321 use flushSync in your AuthProvider when you set your new user/session state like in the thread mentionned by @ilia-luk .

**Comment by syama-aot** [comment]
I've noticed that the issue was not there in version 1.16.1 of @tanstack/react-router, but it was found in the latest release, version 1.56.1. Could you provide a proper solution or let us know if there is a patch release available?

**Comment by joemark0008** [comment]
Got same issue here

 "@tanstack/react-router": "^1.95.3",
 "react": "^19.0.0",
 "react-dom": "^19.0.0",

**Comment by Rick-Phoenix** [comment]
Same here. I am using Router.invalidate() to invalidate router context, but when I hover over links to routes I can see (by console logging the context in these route's beforeload function) that these still have access to the old context. 

Only after a new route is loaded the context gets updated.

**Comment by Rick-Phoenix** [comment]
Yeah in my view router invalidation is not reliable when it comes to auth. In my case I simply force a reload of the page after login/logout with location.href = '/' to make sure the router's state is completely reset.

**Comment by mkarajohn** [comment]
I am facing the same issue, on the latest version. Full repro stackblitz: [here](https://stackblitz.com/edit/tanstack-router-mpppn7py)

I have written some explanatory (hopefully) comments in `AuthProvider`s  `logout` method, as well as in `_authenticated/route.tsx` 

I have also added strategically placed `debugger`s to aid you.

Hope it helps, let me know if you need  more info @schiller-manuel

**Comment by tomasmenezes** [comment]
Facing the same issue. Currently getting around this by reloading post-invalidation:
```typescript
onSignOut: () => {
  router.invalidate();
  router.navigate({ reloadDocument: true });
}
```

"@tanstack/react-router": "^1.130.12"
"react": "^19.1.1"
"react-dom": "^19.1.1"

**Comment by YoavHortman** [comment]
Still experiencing this issue, any progress? seems like tanstack should orchestrate better with react state.......

**Comment by Maadtin** [comment]
Is there some news for this? is `router.invalidate` suposed to re-trigger `load` or `beforeLoad` functions with updated context? happening to me aswell...

**Comment by max-ae** [comment]
I am also running into this issue. `router.invalidate()` does not recompute route context supplied via `beforeLoad`

**Comment by axeloehrli** [comment]
I also have this issue, router.invalidate() is not working for beforeLoad

**Comment by viclafouch** [comment]
Same here.. Impossible to release in production

**Comment by mkarajohn** [comment]
Bump and link to full repro stackblitz: https://github.com/TanStack/router/issues/2072#issuecomment-3124747968

**Comment by max-ae** [comment]
I attempted to fix this in #6598 - unfortunately, I could not get the repro to work locally in a reasonable timeframe, so I hope someone else can validate this fix.
